### PR TITLE
HIVE-26136: Implement UPDATE statements for Iceberg tables

### PR DESCRIPTION
**Pull request by pvary**
### What changes were proposed in this pull request?
Implemented the UPDATE for Iceberg

### Why are the changes needed?
Iceberg update

### Does this PR introduce _any_ user-facing change?
Yes. Update runs

### How was this patch tested?
Unit tests

### CODE REVIEW

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergV2.java [hunk 1]
```diff
@@ -310,6 +310,39 @@ public void testDeleteStatementWithOtherTable() {
         HiveIcebergTestUtils.valueForRow(HiveIcebergStorageHandlerTestUtils.CUSTOMER_SCHEMA, objects), 0);
   }
 
+  @Test
+  public void testUpdateStatementUnpartitioned() {
+    Assume.assumeFalse("Iceberg UPDATEs are only implemented for non-vectorized mode for now", isVectorized);
```
Comment (rbalamohan) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergV2.java:315]: Will reader go via non-vect* path when reading from an updated table?

File: iceberg/iceberg-handler/src/test/queries/positive/update_iceberg_unpartitioned_parquet.q [hunk 2]
```diff
@@ -0,0 +1,26 @@
+drop table if exists tbl_ice;
+create external table tbl_ice(a int, b string, c int) stored by iceberg tblproperties ('format-version'='2');
+
+-- update using simple predicates
+insert into tbl_ice values (1, 'one', 50), (2, 'two', 51), (3, 'three', 52), (4, 'four', 53), (5, 'five', 54), (111, 'one', 55), (333, 'two', 56);
+update tbl_ice set b='Changed' where b in ('one', 'four') or a = 22;
```
Comment (rbalamohan) [inline iceberg/iceberg-handler/src/test/queries/positive/update_iceberg_unpartitioned_parquet.q:6]: It may be good to cover all datatypes supported as well (just guardrails so that it doesn't break later?).

Comment (pvary) [reply iceberg/iceberg-handler/src/test/queries/positive/update_iceberg_unpartitioned_parquet.q:6]: Added unit tests for that.
Thanks for the idea, actually found some issues when the tests were run!

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java [hunk 3]
```diff
@@ -330,61 +331,54 @@ private void commitTable(FileIO io, ExecutorService executor, JobContext jobCont
...
-      Collection<FilesForCommit> results) {
-    RowDelta append = txn.map(Transaction::newRowDelta).orElse(table.newRowDelta());
-    List<DeleteFile> deleteFiles = results.stream().map(FilesForCommit::deleteFiles)
-        .flatMap(Collection::stream).collect(Collectors.toList());
-    deleteFiles.forEach(append::addDeletes);
-    append.commit();
-    LOG.info("Delete commit took {} ms for table: {} with {} delete file(s)",
-        System.currentTimeMillis() - startTime, table, deleteFiles.size());
-    LOG.debug("Added delete files {}", deleteFiles);
+  private void commitWrite(Table table, long startTime, FilesForCommit results) {
```
Comment (lcspinter) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:348]: nit: javadoc

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java [hunk 4]
```diff
@@ -330,61 +331,54 @@ private void commitTable(FileIO io, ExecutorService executor, JobContext jobCont
...
-    } else if (!dataFiles.isEmpty()) {
-      // Appending data files to the table
-      // We only create a new commit if there's something to append
-      AppendFiles append = txn.map(Transaction::newAppend).orElse(table.newAppend());
-      dataFiles.forEach(append::appendFile);
-      append.commit();
-      LOG.info("Append commit took {} ms for table: {} with {} file(s)", System.currentTimeMillis() - startTime, table,
-          dataFiles.size());
-      LOG.debug("Added files {}", dataFiles);
+  private void commitOverwrite(Table table, long startTime, FilesForCommit results) {
```
Comment (lcspinter) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:365]: nit: javadoc

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 5]
```diff
@@ -490,17 +491,30 @@ public List<VirtualColumn> acidVirtualColumns() {
   }
 
   @Override
-  public List<FieldSchema> acidSelectColumns(org.apache.hadoop.hive.ql.metadata.Table table) {
-    // TODO: make it configurable whether we want to include the table columns in the select query
-    // it might make delete writes faster if we don't have to write out the row object
-    return Stream.of(ACID_VIRTUAL_COLS_AS_FIELD_SCHEMA, table.getCols())
-        .flatMap(Collection::stream)
-        .collect(Collectors.toList());
+  public List<FieldSchema> acidSelectColumns(org.apache.hadoop.hive.ql.metadata.Table table, Operation operation) {
```
Comment (lcspinter) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:494]: nit: we could rename this, the other functions and class attributes using the `acid` keyword to something more general, like `virtualMetadataSelectColumns`. We can do this in a separate PR.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergAcidUtil.java [hunk 6]
```diff
@@ -115,20 +115,20 @@ public static PositionDelete<Record> getPositionDelete(Record rec, Record rowDat
    * @return The schema for reading files, extended with metadata columns needed for deletes
    */
   public static Schema createFileReadSchemaForUpdate(List<Types.NestedField> dataCols, Table table) {
-    List<Types.NestedField> cols = Lists.newArrayListWithCapacity(dataCols.size() + SERDE_META_COLS.size());
-    SERDE_META_COLS.forEach((metaCol, index) -> {
+    List<Types.NestedField> cols = Lists.newArrayListWithCapacity(dataCols.size() + FILE_READ_META_COLS.size());
```
Comment (lcspinter) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergAcidUtil.java:118]: nit: Change the name of the class from acid.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 7]
```diff
@@ -3330,6 +3332,23 @@ public static boolean isNonNativeAcidTable(Table table) {
         table.getStorageHandler().supportsAcidOperations() != HiveStorageHandler.AcidSupportType.NONE;
   }
 
+  /**
+   * Returns the virtual columns needed for update queries. For ACID queries it is a single ROW__ID, for non-native
+   * tables the list is provided by the {@link HiveStorageHandler#acidVirtualColumns()}.
+   * @param table The table for which we run the query
+   * @return The list of virtual columns used
+   */
+  public static List<VirtualColumn> getAcidVirtualColumns(Table table) {
```
Comment (lcspinter) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:3341]: nit: rename to getVirtualColumns
